xarray/core/variable.py

```diff
@@ -406,6 +407,96 @@ def __getitem__(self, key):
         return type(self)(dims, values, self._attrs, self._encoding,
                           fastpath=True)
 
+    def _broadcast_indexes(self, key):
+        """
+        Parameters
+        -----------
+        key: One of
+            array
+            a mapping of dimension names to index.
+
+        Returns
+        -------
+        dims: Tuple of strings.
+            Dimension of the resultant variable.
+        indexers: list of integer, array-like, or slice. This is aligned
+            along self.dims.
+        """
+        if not utils.is_dict_like(key):
+            key = {self.dims[0]: key}
+        example_v = None
+        indexes = OrderedDict()
+        for k, v in key.items():
+            if not isinstance(v, (integer_types, slice, Variable)):
+                if not hasattr(key, 'ndim'):  # convert list or tuple
+                    v = np.array(v)
+            if example_v is None and isinstance(v, Variable):
+                example_v = v
+            indexes[k] = v
+
+        # When all the keys are array or integer, slice
```

Comment:
Here's an alternative version that I think will work, at least if the variable's data is a NumPy array (we may need to jump through a few more hoops for dask arrays):
```python
def _broadcast_indexes(self, key):
    key = self._item_key_to_tuple(key)  # key is a tuple
    key = indexing.expanded_indexer(key, self.ndim)  # key is a tuple of full size
    basic_indexing_types = integer_types + (slice,)
    if all(isinstance(k, basic_indexing_types) for k in key):
        return self._broadcast_indexes_basic(key)
    else:
        return self._broadcast_indexes_advanced(key)

def _broadcast_indexes_basic(self, key):
    dims = tuple(dim for k, dim in zip(key, self.dims)
                 if not isinstance(k, integer_types))
    return dims, key

def _broadcast_indexes_advanced(self, key):
    variables = []
    for dim, value in zip(self.dims, key):
        if isinstance(value, slice):
            value = np.arange(*value.indexes(self.sizes[dim])
        # NOTE: this is close to but not quite correct, since we want to
        # handle tuples differently than as_variable and want different
        # error message (not referencing tuples)
        variable = as_variable(value, name=dim)
        variables.append(variable)
    variables = _broadcast_compat_variables(*variables)
    dims = variables[0].dims  # all variables have the same dims
    key = tuple(variable.data for variable in variables)
    return dims, key
```